### PR TITLE
Slightly updated singleinflight code and added test...

### PR DESCRIPTION
**Pull request by asergeyev**
Just to discuss these changes in #250, not for merging yet.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -116,16 +116,9 @@ func (c *Client) Exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err erro
...
 	}
-	r, rtt, err, shared := c.group.Do(m.Question[0].Name+t+cl, func() (*Msg, time.Duration, error) {
+	return c.group.Do(m.Question[0].Name+t+cl, func() (*Msg, time.Duration, error) {
 		return c.exchange(m, a)
 	})
-	if err != nil {
-		return r, rtt, err
-	}
-	if shared {
-		return r.Copy(), rtt, nil
```
Comment (miekg) [inline client.go]: I think this .Copy() is quite important...

File: singleinflight.go [hunk 2]
```diff
@@ -30,28 +32,42 @@ type singleflight struct {
...
+	if ok { // wait in line and copy the result
+		c.dup = true
 		g.Unlock()
-		c.wg.Wait()
-		return c.val, c.rtt, c.err, true
+		c.Wait()
+		if c.val == nil {
+			return c.val, c.rtt, c.err
+		}
+		return c.val.Copy(), c.rtt, c.err
```
Comment (miekg) [inline singleinflight.go]: ah, you copy here...

**Comment by miekg** [comment]
What is it you actually improving here?

**Comment by asergeyev** [comment]
I wasn't really planning to change singleinflight (at least yet), I was just poking at things and it could be rolled back.. Focus was on creating a  test... but now I see that it failed anyway and wondering why it happens...

**Pull request closed without merging**